File: 2025-season6/3주차/최영권/AddTwoNumbers.java
package LinkedList;

import LinkedList.Node.ListNode;

import java.math.BigInteger;

public class AddTwoNumbers {

    public static void main(String[] args) {
        AddTwoNumbers addTwoNumbers = new AddTwoNumbers();
        ListNode l1 = new ListNode(1, new ListNode(2, new ListNode(3)));
        ListNode l2 = new ListNode(2, new ListNode(1, new ListNode(3, new ListNode(9))));
//        ListNode l2 = new ListNode();
        ListNode listNode = addTwoNumbers.addTwoNumbers2(l1, l2);
        System.out.println(listNode);
    }

    /*
     * 시간복잡도 : O(max(m, n))
     * - m: l1의 길이, n: l2의 길이
     * - 더 긴 연결리스트의 길이에 비례
     */
    public ListNode addTwoNumbers2(ListNode l1, ListNode l2) {
        int carry = 0;
        ListNode head = new ListNode();
        ListNode listNode = head;

        while (l1 != null || l2 != null || carry > 0) {
            int val1 = (l1 != null) ? l1.val : 0;
            int val2 = (l2 != null) ? l2.val : 0;

            int sum = val1 + val2 + carry;
            carry = sum / 10;
            listNode.val = sum % 10;

            // 다음 노드가 필요한 경우에만 생성
            if (l1 != null || l2 != null || carry > 0) {
                if (l1 != null) l1 = l1.next;
                if (l2 != null) l2 = l2.next;

                // 아직 처리할 것이 남아있다면 다음 노드 생성
                if (l1 != null || l2 != null || carry > 0) {
                    listNode.next = new ListNode();
                    listNode = listNode.next;
                }
            }
        }

        return head;
    }

    // solution 1
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        String l1Str = reverse(l1);
        String l2Str = reverse(l2);

        BigInteger l = new BigInteger(l1Str);
        BigInteger l3 = new BigInteger(l2Str);
        BigInteger add = l.add(l3);
        String string = add.toString();
        char[] charArray = string.toCharArray();
        ListNode head = new ListNode();
        ListNode listNode = head;
        for (int i = charArray.length - 1; i >= 0; i--) {
            listNode.val = Character.getNumericValue(charArray[i]);
            if (i != 0) {
                listNode.next = new ListNode();
                listNode = listNode.next;
            }
        }
        return head;
    }

    // 1,2,3,4 -> 4321
    private String reverse(ListNode l1) {
        ListNode prev = null;
        ListNode curr = l1;
        StringBuilder result = new StringBuilder();
        while (curr != null) {
            result.insert(0, curr.val);
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return result.toString();
    }


}

File: 2025-season6/3주차/최치환/add-two-numbers.kt
class Solution {
    // 전체 시간복잡도: O(1) + O(1) + O(max(m, n)) + O(1) = O(max(m, n))
    // 전체 공간복잡도:O(max(m, n))

    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        // O(1)
        var sum = (l1?.`val` ?: 0) + (l2?.`val` ?: 0)
        var temp: Int = sum / 10

        // O(1)
        val result: ListNode? = ListNode(sum % 10)
        var resultHead = result
        var l1Head = l1?.next
        var l2Head = l2?.next

        // O(max(m, n)) : 두 리스트 전체 길이만큼 반복 (m, n은 각 리스트의 길이)
        while (l1Head?.`val` != null || l2Head?.`val` != null) {
            // O(1) : 각 자릿수의 합 계산 및 노드 연결
            sum = (l1Head?.`val` ?: 0) + (l2Head?.`val` ?: 0) + temp
            resultHead?.next = ListNode((sum % 10))
            resultHead = resultHead?.next
            temp = sum / 10
            l1Head = l1Head?.next
            l2Head = l2Head?.next
        }

        // O(1) : 마지막 자리 올림 처리
        if (temp != 0) {
            resultHead?.next = ListNode(temp)
        }

        return result
    }
}